config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml

```
@@ -1,23 +1,17 @@
-
----
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
-  annotations:
-    controller-gen.kubebuilder.io/version: v0.6.2
-  creationTimestamp: null
   name: kongclusterplugins.configuration.konghq.com
 spec:
   group: configuration.konghq.com
+  scope: Cluster
```

```suggestion
  scope: Namespaced
```
CRD themselves are non-namespaced and are available to all namespaces

---

`spec.scope` does not indicate the namespace of the CRD type here, it indicates the scope of the `KongClusterPlugin` type. This needs to remain `scope: Cluster`.

---

Right, but CRs you create from a CRD aren't; _those_ are _either_ namespaced or cluster-wide, and the `scope` field of the CRD indicates which type the CRs are. KongClusterPlugins are the one CR we have that is cluster-wide (whereas KongPlugins are namespaced), so this is correct as-is and should not be changed.

---

well, I am exactly suggesting "Namespaced" CR, all Kong resource should be located into Kong namespace (kong, kong-system etc). Thats the point unless we have technical blocker atm.

---

It seems we may have a misunderstanding here and I want to make sure we get that resolved:

> well, I am exactly suggesting "Namespaced" CR, all Kong resource(s) should be located into Kong namespace (kong, kong-system etc). Thats the point unless we have technical blocker atm.

This is explicitly and historically not the case for the `KongClusterPlugin` type.

It sounds to me like maybe what you're advocating for is an intentional change of the scope, but the problem with that is this _would be a backwards incompatible breaking change_ and we would need to develop new tools and upgrade strategies to support it.

I would recommend @ccfishk if you have some reasons you would like to see this changed that you bring that up as a separate issue for after 2.0, and include your reasoning because at least now we couldn't make this change without breaking things (and having to make significant changes internally) and I also personally don't see or understand the impetus to do it.

---

Well, @rainest, and shane,  after more thought, "Namespaced" scope still have more advantage even in the deploying same CRD repeatedly cases - Namespace gives user the flexibility of different version in the same cluster, which justified its popularity. "Cluster" scope lost different CR version.

---

> we should revert all back to "Namespaced" scope.

The word "revert" indicates that you want to change something _back_ to a previous state.

Again, please see this link:

https://github.com/Kong/kubernetes-ingress-controller/blob/v1.3.2/deploy/manifests/base/custom-types.yaml#L86

`scope: Cluster` **is the previous state**.

---

I understand the motivation of "Cluster" scope, just illustrating the pros and cons between the two.

---

Sounds good, but we shouldn't try to change this now, in this scope, right before launch. I mentioned earlier in this thread that it would be good if you could create a new issue so that we can look into changing this later down the road, are you open to doing that and considering this thread "resolved" based on that follow up?

---

Sorry, I am really confused. 
I believe basically this is - since the need is from past customer complain.

---

I'm confused as well, can you share with me the customer complaint?

---

here it is, Shane, copy/paste from  https://github.com/Kong/kubernetes-ingress-controller/pull/1763#discussion_r700567222
"deploying same CR repeatedly in different namespaces" if "Namespace" scoped.

---

That doesn't seem to be a "customer complaint", but rather it seems to be your own supposition that there's something to improve. I do not see sufficient cause to make this breaking change as a part of this PR.

---

No. I throw this question when I see the back and force between "Cluster" and "Namespaced" in different PR(s), travis brought the complaining use cases, I see the advantages and cons. Kindly check with him since he has a whole story. 👍

---

My explanation earlier was to try and lay out why KongClusterPlugin _does_ have cluster scope, not that we should change it. If there were no cluster-scoped plugin configuration option, users would have no way to use the same plugin configuration for Ingresses in different namespaces and _would_ have to copy their KongPlugins to each namespace each time they wanted to use the same configuration across namespaces.

To try and pose this a different way, if we were to make this change, how would KongClusterPlugin and KongPlugin differ? For quick reference:
https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/deploy/manifests/base/custom-types.yaml#L81-L86
https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/deploy/manifests/base/custom-types.yaml#L4-L9

---

Given the approval on the PR, it seems like this is now resolved. Let me know if that is not the case :+1: